File: Advent-of-Code-2021/day15.ts
import * as utils from "./utils.ts";

const grid: number[][] = utils.readDay(15).map((line) =>
    [...line].map(Number)
);

function solve(grid: number[][]) {
    const distances = grid.map((row) => row.map(() => Infinity));
    distances[0][0] = 0;
    const height = grid.length;
    const width = grid[0].length;

    const frontier: [number, number][] = [[0, 0]];
    while (frontier.length) {
        // One day let's write a priority queue, and maybe use A*
        const [y, x] = frontier
            .sort(([y1, x1], [y2, x2]) => distances[y2][x2] - distances[y1][x1])
            .pop()!;
        for (const [ny, nx] of utils.neighbors(y, x, height, width)) {
            if (distances[ny][nx] > grid[ny][nx] + distances[y][x]) {
                distances[ny][nx] = grid[ny][nx] + distances[y][x];
                frontier.push([ny, nx]);
            }
        }
    }
    return distances[height - 1][width - 1];
}
console.log(solve(grid));

const bigGrid = utils.range(5)
    .flatMap((i) => grid.map(row => utils.range(5)
        .flatMap((j) => row.map((n) => (n + i + j - 1) % 9 + 1))));
console.log(solve(bigGrid));

File: Advent-of-Code-2021/day15.py
import heapq

grid = [[int(c) for c in row.strip()] for row in open("day15.txt").readlines()]


def solve(grid: list[list[int]]) -> int:
    distances = [[2 ** 31 for _ in row] for row in grid]
    distances[0][0] = 0
    HEIGHT = len(grid)
    WIDTH = len(grid[0])

    def neighbors(y: int, x: int) -> list[tuple[int, int]]:
        return [
            (y + dy, x + dx)
            for dy, dx in ((-1, 0), (1, 0), (0, -1), (0, 1))
            if 0 <= y + dy < HEIGHT and 0 <= x + dx < WIDTH
        ]

    frontier = [(0, (0, 0))]
    while frontier:
        cost, (y, x) = heapq.heappop(frontier)
        for (ny, nx) in neighbors(y, x):
            if distances[ny][nx] > grid[ny][nx] + cost:
                new_cost = grid[ny][nx] + cost
                distances[ny][nx] = new_cost
                heapq.heappush(frontier, (new_cost, (ny, nx)))
    return distances[HEIGHT - 1][WIDTH - 1]


print(solve(grid))

big_grid = [
    [(c + i + j - 1) % 9 + 1 for j in range(5) for c in row]
    for i in range(5)
    for row in grid
]
print(solve(big_grid))
